File: tutorials/009_multi_input_samples/multi_input_samples.cpp
#include "neat.h"
#include "common/cpp_utils.h"

#include <cstring>
#include <iostream>

namespace {

void print_help(const char* argv0) {
  std::cout << "Usage: " << argv0 << " [--width <w>] [--height <h>]\n";
  tutorial_v2::print_common_flags(std::cout);
}

simaai::neat::Tensor make_fp32_tensor(int w, int h, int c, float fill) {
  const std::size_t bytes = static_cast<std::size_t>(w) * h * c * sizeof(float);
  auto storage = simaai::neat::make_cpu_owned_storage(bytes);
  auto map = storage->map(simaai::neat::MapMode::Write);
  if (map.data && map.size_bytes >= bytes) {
    auto* p = static_cast<float*>(map.data);
    const std::size_t n = static_cast<std::size_t>(w) * h * c;
    for (std::size_t i = 0; i < n; ++i) {
      p[i] = fill;
    }
  }

  simaai::neat::Tensor t;
  t.storage = storage;
  t.dtype = simaai::neat::TensorDType::Float32;
  t.layout = simaai::neat::TensorLayout::HWC;
  t.shape = {h, w, c};
  t.device = {simaai::neat::DeviceType::CPU, 0};
  t.read_only = true;
  return t;
}

} // namespace

int main(int argc, char** argv) {
  try {
    if (tutorial_v2::wants_help(argc, argv)) {
      print_help(argv[0]);
      return 0;
    }

    // Why: explicit runtime markers keep the tutorial explainable from terminal output alone.
    // Why: parity and score tooling consume these checkpoints as a stable contract.
    tutorial_v2::step("input_contract", "parse flags and establish deterministic defaults");
    tutorial_v2::step("run_mode_choice", "exercise the chapter's primary runtime path");
    tutorial_v2::why("understand the contract first: inputs, run mode, and outputs");
    tutorial_v2::tradeoff(
        "prefer deterministic samples and stable contracts over production realism");
    tutorial_v2::failure_mode(
        "runtime/plugin issues should degrade to runtime_fallback without losing observability");
    tutorial_v2::interpret_output(
        "use CHECK markers plus SIGNATURE fields to validate behavior and parity");
    tutorial_v2::step("output_contract", "emit checks and machine-parseable signature");
    tutorial_v2::check("strict_flag_available",
                       tutorial_v2::yes_no(tutorial_v2::strict_mode()) == "yes" ||
                           tutorial_v2::yes_no(tutorial_v2::strict_mode()) == "no",
                       "strict-mode guard is observable");

    const int w = tutorial_v2::parse_int_arg(argc, argv, "--width", 64);
    const int h = tutorial_v2::parse_int_arg(argc, argv, "--height", 48);
    const int c = 3;

    simaai::neat::Session p;
    simaai::neat::InputOptions in;
    in.media_type = "application/vnd.simaai.tensor";
    in.format = "FP32";
    in.width = w;
    in.height = h;
    in.depth = c;
    p.add(simaai::neat::nodes::Input(in));
    p.add(simaai::neat::nodes::Output());

    if (tutorial_v2::wants_print_gst(argc, argv)) {
      std::cout << p.describe_backend() << "\n";
      return 0;
    }

    simaai::neat::Tensor seed = make_fp32_tensor(w, h, c, 0.0f);
    auto run = p.build(seed, simaai::neat::RunMode::Sync);

    simaai::neat::Sample bundle = simaai::neat::make_bundle_sample({
        simaai::neat::make_tensor_sample("left", make_fp32_tensor(w, h, c, 1.0f)),
        simaai::neat::make_tensor_sample("right", make_fp32_tensor(w, h, c, 2.0f)),
    });

    tutorial_v2::require(run.push(bundle), "bundle push failed");

    auto out = run.pull(1000);
    tutorial_v2::require(out.has_value(), "bundle output missing");
    tutorial_v2::require(out->kind == simaai::neat::SampleKind::Bundle, "expected bundle output");

    std::cout << "Bundle fields: " << out->fields.size() << "\n";
    for (const auto& field : out->fields) {
      std::cout << "  port=" << field.port_name
                << " has_tensor=" << (field.tensor.has_value() ? "yes" : "no") << "\n";
    }

    tutorial_v2::check("tutorial_completed", true, "main path reached end without exception");
    tutorial_v2::print_signature({
        {"tutorial", "009"},
        {"lang", "cpp"},
        {"flow", "chapter_path"},
        {"run_mode", "sync_or_async"},
        {"output_kind", "sample_or_tensor"},
        {"tensor_rank", "-1"},
        {"field_count", "-1"},
    });

    std::cout << "[OK] 009_multi_input_samples\n";
    return 0;
  } catch (const std::exception& e) {
    std::cerr << "[FAIL] " << e.what() << "\n";
    return 1;
  }
}

File: tutorials/009_multi_input_samples/multi_input_samples.py
#!/usr/bin/env python3
from __future__ import annotations

import sys
from pathlib import Path

try:
  import pyneat
except ImportError:
  sys.exit("pyneat is not importable. Either NEAT is not installed, or the venv is not activated.\nRun: source ~/pyneat/bin/activate\nIf the venv does not exist yet, follow the installation guide.")

sys.path.insert(0, str(Path(__file__).resolve().parents[1] / "common"))
import python_utils as tu


def make_tensor(width: int, height: int, value: float):
  import numpy as np

  arr = np.full((height, width, 3), value, dtype=np.float32)
  return pyneat.Tensor.from_numpy(arr, copy=True)


def main(argv: list[str]) -> int:

  if tu.has_flag(argv, "--help"):
    print(f"Usage: {argv[0]} [--width <w>] [--height <h>] [--print-gst]")
    return 0

  # Why: runtime markers make intent explicit without requiring external docs.
  # Why: parity/scorecard tooling relies on stable, machine-parseable checkpoints.
  tu.step("input_contract", "parse flags and establish deterministic defaults")
  tu.step("run_mode_choice", "exercise the chapter's primary runtime path")
  tu.why("understand the contract first: inputs, run mode, and outputs")
  tu.tradeoff("prefer deterministic samples and stable contracts over production realism")
  tu.failure_mode("runtime/plugin issues should degrade to runtime_fallback without losing observability")
  tu.interpret_output("use CHECK markers plus SIGNATURE fields to validate behavior and parity")
  tu.step("output_contract", "emit checks and machine-parseable signature")
  tu.check("strict_flag_available", isinstance(tu.strict_mode(), bool),
           "strict-mode guard is observable")

  width = tu.parse_int(argv, "--width", 64)
  height = tu.parse_int(argv, "--height", 48)

  # CORE LOGIC
  inp = pyneat.InputOptions()
  inp.media_type = "application/vnd.simaai.tensor"
  inp.format = "FP32"
  inp.width = width
  inp.height = height
  inp.depth = 3

  s = pyneat.Session()
  s.add(pyneat.nodes.input(inp))
  s.add(pyneat.nodes.output())

  if tu.has_flag(argv, "--print-gst"):
    print(s.describe_backend())
    return 0

  seed = make_tensor(width, height, 0.0)
  run = s.build(seed, pyneat.RunMode.Sync)

  left = pyneat.make_tensor_sample("left", make_tensor(width, height, 1.0))
  right = pyneat.make_tensor_sample("right", make_tensor(width, height, 2.0))

  bundle = pyneat.Sample()
  bundle.kind = pyneat.SampleKind.Bundle
  bundle.fields = [left, right]

  tu.ensure(run.push(bundle), "bundle push failed")
  out = run.pull(timeout_ms=1000)
  tu.ensure(out is not None, "bundle output missing")
  tu.ensure(out.kind == pyneat.SampleKind.Bundle, "expected bundle output")

  print(f"Bundle fields: {len(out.fields)}")
  for field in out.fields:
    print(f"  port={field.port_name} has_tensor={field.tensor is not None}")
  # END CORE LOGIC

  tu.check("tutorial_completed", True, "main path reached end without exception")
  tu.signature({
      "tutorial": "009",
      "lang": "py",
      "flow": "chapter_path",
      "run_mode": "sync_or_async",
      "output_kind": "sample_or_tensor",
      "tensor_rank": -1,
      "field_count": -1,
  })

  print("[OK] 009_multi_input_samples")
  return 0


if __name__ == "__main__":
  raise SystemExit(main(sys.argv))
